locator/tablets.cc

```diff
@@ -245,9 +246,42 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
     }
 }
 
-host_id tablet_map::get_primary_replica(tablet_id id) const {
+const tablet_replica& tablet_map::sort_and_get_primary_replica(tablet_replica_set& replicas, tablet_id id) {
```

Comment:
Passing tablet_replica_set& which allows modifying the replicas at the caller side looks dangerous. People might misuse it to sort the info.replicas unintentionally. 

How about:

tablet_replica tablet_map::get_primary_replica(const tablet_replica_set& replicas, tablet_id id). 

This is enough even for the unitest below.

Reply:
And let repair di the filtering for the dc?
That's possible.  Though a bit of an overkill since this is a private method by design and not intended for public use.
Also, its name suggests it's sorting the input vector so it shouldn't come as a surprise to the caller.

Reply:
I meant we keep the exiting 

tablet_replica tablet_map::get_primary_replica_within_dc(tablet_id id, const topology& topo, const sstring& dc) tablet_replica tablet_map::get_primary_replica(tablet_id id) const

in addition, we have

tablet_replica tablet_map::get_primary_replica( tablet_id id, const tablet_replica_set& replicas)

which can be used to get tablet_replica with the given replicas provided. 

The filter stuff is still done by the  tablet_replica tablet_map::get_primary_replica_within_dc.

Reply:
This will add yet another copy of the replicas within the dc.
To save it, the low-level get_primary_replica function can receive also a predicate function to filter the `const tablet_replica_set& replica` while copying it, before sorting the copy in place.